lib/srv/alpnproxy/proxy.go

```diff
@@ -101,6 +101,39 @@ func MatchByALPNPrefix(prefix string) MatchFunc {
 	}
 }
 
+// MatchMySQLConn return a pre-process function for MySQL connections that tries to extract MySQL server version
+// from incoming connection.
+func MatchMySQLConn(fn func(ctx context.Context, conn net.Conn) error) HandlerFuncWithInfo {
+	return func(ctx context.Context, conn net.Conn, info ConnectionInfo) error {
+		for _, alpn := range info.ALPN {
+			if !strings.HasPrefix(alpn, string(common.ProtocolMySQL)) || alpn == string(common.ProtocolMySQL) {
+				continue
+			}
+
+			// prefix and '-' character
+			const mysqlVerStart = len(common.ProtocolMySQL) + 1
+			// Check if the name contains at least one character
+			// 2 = 1 ('-' char) + 1 (at least one character of version string)
+			if len(alpn) <= mysqlVerStart+1 || alpn[mysqlVerStart-1] != '-' {
+				continue
+			}
```

Comment:
I think we can remove some checks here

```suggestion
			mysqlVerPrefix := string(common.ProtocolMySQL)+"-"
			mysqlVerStart := len(mysqlVerPrefix)
			if !strings.HasPrefix(alpn, mysqlVerPrefix)  || len(alpn) == mysqlVerStart {
				continue
			}
```

Reply:
Thanks. I've changed your suggestion a bit. I think now the code is easier to understand.